File: iphone/SpeedConsumerApp/TrafficSignCountryPackRegistry.swift
import Foundation
import CryptoKit

/// Discovery never activates a runtime. The installer must verify the referenced
/// delivery envelope, all files, licenses and acceptance evidence independently.
struct TrafficSignCountryPackRegistry: Decodable, Sendable {
    struct ArtifactReference: Decodable, Sendable {
        let url: String
        let sha256: String
        let sizeBytes: Int
    }
    struct Compatibility: Decodable, Sendable {
        struct Platform: Decodable, Sendable {
            let platform: String
            let minimumRuntime: String
        }
        let minimumAppVersion: String
        let maximumAppVersion: String
        let taxonomyVersion: String
        let preprocessingVersion: String
        let platforms: [Platform]
    }
    struct Pack: Decodable, Sendable {
        let packId: String
        let version: Int
        let countries: [String]
        let rollout: String
        let manifest: ArtifactReference?
        let compatibility: Compatibility
        let calibrated: Bool
        let reason: String
        let supersedes: [Int]
        let rollbackVersion: Int?
    }
    struct Decision: Equatable {
        let state: String
        let packId: String?
        let version: Int?
        // Discovery is never permission for live speed overrides.
        let overrideEligible = false
    }
    let schemaVersion: Int
    let generation: Int
    let environment: String
    let expiresAt: Int
    let packs: [Pack]

    /// Trust pins must be shipped/reviewed out of band, never read from the
    /// downloaded index. No production remote pin or endpoint is enabled yet.
    static func decodeTrusted(_ data: Data, trustedSHA256: Set<String>, minimumGeneration: Int = 1, allowFixtures: Bool = false) throws -> Self {
        guard data.count <= 1_048_576 else { throw DiscoveryError.invalidRegistry }
        let digest = SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
        guard trustedSHA256.contains(digest) else { throw DiscoveryError.untrustedRegistry }
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        let result = try decoder.decode(Self.self, from: data)
        guard result.schemaVersion == 1, (1...Int(Int32.max)).contains(result.generation),
              result.generation >= minimumGeneration, result.expiresAt > 0, result.expiresAt <= 4_102_444_800,
              ["fixture", "production"].contains(result.environment),
              allowFixtures || result.environment == "production", result.packs.count <= 256,
              Set(result.packs.map { "\($0.packId):\($0.version)" }).count == result.packs.count else { throw DiscoveryError.invalidRegistry }
        for pack in result.packs {
            let c = pack.compatibility
            guard pack.packId.range(of: "^[a-z0-9][a-z0-9._-]{0,95}$", options: .regularExpression) != nil,
                  (1...Int(Int32.max)).contains(pack.version), !pack.countries.isEmpty, pack.countries.count <= 16,
                  Set(pack.countries).count == pack.countries.count,
                  pack.countries.allSatisfy({ $0.range(of: "^[A-Z]{2}$", options: .regularExpression) != nil }),
                  ["evaluation", "shadow", "canary", "available", "withdrawn"].contains(pack.rollout),
                  let minimum = numericVersion(c.minimumAppVersion), let maximum = numericVersion(c.maximumAppVersion), minimum <= maximum,
                  !c.taxonomyVersion.isEmpty, !c.preprocessingVersion.isEmpty, !pack.reason.isEmpty,
                  !c.platforms.isEmpty, c.platforms.count <= 2,
                  Set(c.platforms.map(\.platform)).count == c.platforms.count,
                  c.platforms.allSatisfy({ ["ios", "android"].contains($0.platform) && numericVersion($0.minimumRuntime) != nil }),
                  Set(pack.supersedes).count == pack.supersedes.count,
                  pack.supersedes.allSatisfy({ $0 > 0 && $0 < pack.version }),
                  pack.rollbackVersion.map({ $0 > 0 && $0 < pack.version }) ?? true else { throw DiscoveryError.invalidRegistry }
            if let ref = pack.manifest {
                guard (1...1_048_576).contains(ref.sizeBytes),
                      ref.sha256.range(of: "^[a-f0-9]{64}$", options: .regularExpression) != nil,
                      ref.url.range(of: "^https://[a-z0-9.-]+/[A-Za-z0-9/_.-]+$", options: .regularExpression) != nil,
                      ref.url.count <= 2048, ref.url.contains("/\(ref.sha256)/"),
                      !ref.url.split(separator: "/").contains("..") else { throw DiscoveryError.invalidRegistry }
            }
        }
        return result
    }

    static func bundled(_ bundle: Bundle = .main) -> Self? {
        guard let url = bundle.url(forResource: "country-pack-registry-v1", withExtension: "json"),
              let data = try? Data(contentsOf: url) else { return nil }
        // The signed application bundle is the trust root for this inventory.
        let digest = SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
        return try? decodeTrusted(data, trustedSHA256: [digest])
    }

    func decision(country: String?, platform: String, appVersion: String, runtimeVersion: String,
                  taxonomy: String = "tsr-semantic-v1", preprocessing: String = "vision-scale-fit-rgb-v1", now: Int) -> Decision {
        func result(_ state: String, _ pack: Pack? = nil) -> Decision { Decision(state: state, packId: pack?.packId, version: pack?.version) }
        guard now >= 0, now < expiresAt else { return result("registry_expired") }
        guard let country else { return result("country_unresolved") }
        let candidates = packs.filter { $0.countries.contains(country) }
        // A country cannot silently choose between competing independently named packs.
        guard Set(candidates.map(\.packId)).count <= 1 else { return result("ambiguous_pack") }
        guard let pack = candidates.max(by: { $0.version < $1.version }) else { return result("unavailable") }
        if pack.rollout == "withdrawn" { return result("withdrawn", pack) }
        guard let app = Self.numericVersion(appVersion), let runtime = Self.numericVersion(runtimeVersion),
              let minimum = Self.numericVersion(pack.compatibility.minimumAppVersion),
              let maximum = Self.numericVersion(pack.compatibility.maximumAppVersion), app >= minimum, app <= maximum,
              pack.compatibility.taxonomyVersion == taxonomy, pack.compatibility.preprocessingVersion == preprocessing,
              let target = pack.compatibility.platforms.first(where: { $0.platform == platform }),
              let minimumRuntime = Self.numericVersion(target.minimumRuntime), runtime >= minimumRuntime else { return result("incompatible", pack) }
        guard pack.manifest != nil else { return result("unavailable", pack) }
        if environment == "fixture" { return result("fixture_only", pack) }
        if !pack.calibrated || ["evaluation", "shadow"].contains(pack.rollout) { return result("shadow", pack) }
        if pack.rollout == "canary" { return result("canary_not_enrolled", pack) }
        return result("downloadable", pack)
    }

    // 1.1 and 1.1.0 are equal for the installed app/OS; registry versions are
    // validated separately by the release schema. Suffixes are never stripped.
    private static func numericVersion(_ raw: String) -> Int64? {
        guard raw.range(of: "^[0-9]{1,6}(\\.[0-9]{1,6}){0,2}$", options: .regularExpression) != nil else { return nil }
        var parts = raw.split(separator: ".").compactMap { Int64($0) }
        while parts.count < 3 { parts.append(0) }
        return parts[0] * 1_000_000_000_000 + parts[1] * 1_000_000 + parts[2]
    }
}

File: android/app/src/main/java/de/youspeed/android/alpha/TrafficSignCountryPackRegistry.kt
package de.youspeed.android.alpha

import java.security.MessageDigest
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.long

/** Discovery only. Installation and reviewed runtime activation remain separate gates. */
data class TrafficSignCountryPackRegistry(
    val generation: Int, val environment: String, val expiresAt: Long, val packs: List<Pack>,
) {
    data class ArtifactReference(val url: String, val sha256: String, val sizeBytes: Int)
    data class Platform(val platform: String, val minimumRuntime: String)
    data class Compatibility(
        val minimumAppVersion: String, val maximumAppVersion: String,
        val taxonomyVersion: String, val preprocessingVersion: String, val platforms: List<Platform>,
    )
    data class Pack(
        val packId: String, val version: Int, val countries: List<String>, val rollout: String,
        val manifest: ArtifactReference?, val compatibility: Compatibility, val calibrated: Boolean,
        val reason: String, val supersedes: List<Int>, val rollbackVersion: Int?,
    )
    data class Decision(val state: String, val packId: String? = null, val version: Int? = null) {
        val overrideEligible: Boolean get() = false
    }

    fun decision(
        country: String?, platform: String, appVersion: String, runtimeVersion: String,
        taxonomy: String = "tsr-semantic-v1", preprocessing: String = "vision-scale-fit-rgb-v1", now: Long,
    ): Decision {
        fun result(state: String, pack: Pack? = null) = Decision(state, pack?.packId, pack?.version)
        if (now < 0 || now >= expiresAt) return result("registry_expired")
        if (country == null) return result("country_unresolved")
        val candidates = packs.filter { country in it.countries }
        if (candidates.map { it.packId }.toSet().size > 1) return result("ambiguous_pack")
        val pack = candidates.maxByOrNull { it.version } ?: return result("unavailable")
        if (pack.rollout == "withdrawn") return result("withdrawn", pack)
        val c = pack.compatibility
        val app = numericVersion(appVersion); val runtime = numericVersion(runtimeVersion)
        val minimum = numericVersion(c.minimumAppVersion); val maximum = numericVersion(c.maximumAppVersion)
        val target = c.platforms.firstOrNull { it.platform == platform }
        val minimumRuntime = target?.let { numericVersion(it.minimumRuntime) }
        if (app == null || runtime == null || minimum == null || maximum == null || minimumRuntime == null ||
            app !in minimum..maximum || runtime < minimumRuntime || c.taxonomyVersion != taxonomy ||
            c.preprocessingVersion != preprocessing) return result("incompatible", pack)
        if (pack.manifest == null) return result("unavailable", pack)
        if (environment == "fixture") return result("fixture_only", pack)
        if (!pack.calibrated || pack.rollout in setOf("evaluation", "shadow")) return result("shadow", pack)
        if (pack.rollout == "canary") return result("canary_not_enrolled", pack)
        return result("downloadable", pack)
    }

    companion object {
        fun sha256(bytes: ByteArray): String = MessageDigest.getInstance("SHA-256").digest(bytes).joinToString("") { "%02x".format(it) }

        /** Only for bytes from the signed APK, never for a downloaded index. */
        fun decodeBundled(bytes: ByteArray): TrafficSignCountryPackRegistry = decodeTrusted(bytes, setOf(sha256(bytes)))

        fun decodeTrusted(bytes: ByteArray, trustedSHA256: Set<String>, minimumGeneration: Int = 1, allowFixtures: Boolean = false): TrafficSignCountryPackRegistry {
            require(bytes.size <= 1_048_576)
            require(sha256(bytes) in trustedSHA256) { "Untrusted registry" }
            val root = Json.parseToJsonElement(bytes.toString(Charsets.UTF_8)).jsonObject
            require(root.getValue("schema_version").jsonPrimitive.int == 1)
            val generation = root.getValue("generation").jsonPrimitive.int
            val environment = root.text("environment")
            val expiresAt = root.getValue("expires_at").jsonPrimitive.long
            require(generation >= maxOf(1, minimumGeneration) && expiresAt in 1..4_102_444_800L)
            require(environment in setOf("fixture", "production") && (allowFixtures || environment == "production"))
            val packs = root.getValue("packs").jsonArray.map { element ->
                val item = element.jsonObject
                val c = item.getValue("compatibility").jsonObject
                val compatibility = Compatibility(c.text("minimum_app_version"), c.text("maximum_app_version"),
                    c.text("taxonomy_version"), c.text("preprocessing_version"), c.getValue("platforms").jsonArray.map {
                        Platform(it.jsonObject.text("platform"), it.jsonObject.text("minimum_runtime"))
                    })
                val manifest = item.getValue("manifest").takeUnless { it == JsonNull }?.jsonObject?.let {
                    ArtifactReference(it.text("url"), it.text("sha256"), it.getValue("size_bytes").jsonPrimitive.int)
                }
                Pack(item.text("pack_id"), item.getValue("version").jsonPrimitive.int,
                    item.getValue("countries").jsonArray.map { it.jsonPrimitive.content }, item.text("rollout"),
                    manifest, compatibility, item.getValue("calibrated").jsonPrimitive.boolean, item.text("reason"),
                    item.getValue("supersedes").jsonArray.map { it.jsonPrimitive.int },
                    item.getValue("rollback_version").takeUnless { it == JsonNull }?.jsonPrimitive?.int,
                ).also { pack ->
                    require(Regex("[a-z0-9][a-z0-9._-]{0,95}").matches(pack.packId) && pack.version > 0)
                    require(pack.countries.size in 1..16 && pack.countries.toSet().size == pack.countries.size &&
                        pack.countries.all { Regex("[A-Z]{2}").matches(it) })
                    require(pack.rollout in setOf("evaluation", "shadow", "canary", "available", "withdrawn"))
                    val minimum = requireNotNull(numericVersion(compatibility.minimumAppVersion))
                    val maximum = requireNotNull(numericVersion(compatibility.maximumAppVersion))
                    require(minimum <= maximum && compatibility.taxonomyVersion.isNotEmpty() && compatibility.preprocessingVersion.isNotEmpty())
                    require(pack.reason.isNotEmpty() && compatibility.platforms.size in 1..2 &&
                        compatibility.platforms.map { it.platform }.toSet().size == compatibility.platforms.size)
                    require(compatibility.platforms.all { it.platform in setOf("ios", "android") && numericVersion(it.minimumRuntime) != null })
                    require(pack.supersedes.toSet().size == pack.supersedes.size && pack.supersedes.all { it in 1 until pack.version })
                    require(pack.rollbackVersion == null || pack.rollbackVersion in 1 until pack.version)
                    manifest?.let { ref ->
                        require(ref.sizeBytes in 1..1_048_576 && Regex("[a-f0-9]{64}").matches(ref.sha256))
                        require(ref.url.length <= 2048 && Regex("https://[a-z0-9.-]+/[A-Za-z0-9/_.-]+").matches(ref.url) &&
                            "/${ref.sha256}/" in ref.url && ".." !in ref.url.split('/'))
                    }
                }
            }
            require(packs.size <= 256 && packs.map { it.packId to it.version }.toSet().size == packs.size)
            return TrafficSignCountryPackRegistry(generation, environment, expiresAt, packs)
        }

        private fun JsonObject.text(key: String): String {
            val value = getValue(key).jsonPrimitive
            require(value.isString)
            return value.content
        }

        private fun numericVersion(raw: String): Long? {
            if (!Regex("[0-9]{1,6}(\\.[0-9]{1,6}){0,2}").matches(raw)) return null
            val parts = raw.split('.').map { it.toLong() } + listOf(0L, 0L)
            return parts[0] * 1_000_000_000_000L + parts[1] * 1_000_000L + parts[2]
        }
    }
}
